[emscripten] Web build

ololoken opened the pull request:
I'm working on wasm port.
Demo is available here https://turch.in/rttr/index.html
<img width="1472" alt="image" src="https://github.com/user-attachments/assets/95dd8346-4378-46fc-8eb2-45cf8df0b297" />
Right now it's not playable. As emscripten do not support socket listen in web environment. 
And I need good advice: how do you (settlers freaks) see implementation of local connection without sockets?

ololoken commented:
@Flamefire well, motivation is to get original fill and look. First I ported [widelands](https://github.com/widelands), playable version is here https://turch.in/widelands/ . But it's a bit different game... Besides, it's cool to have version of the game which doesn't require download and install, it just works right now and here in browser.
Yes there are performance issues, wasm is executed in 32bit vm with jit, but I think it's not a big problem as machines gets faster(slower than before, but steady) and wasm vm soon may get fat performance boost with upcoming updates. 
I will go through comments later and add fixes. First I want to consider workaround for network problem. I agree that adding of custom logic for local game is bad idea in matter of maintenatability.  There few options I can try: it's time to write own websocket-tcp/udp proxy, hack libwebsocket.js to make virtual peer connections for localhost.

Flamefire commented on libs/s25main/ogl/glArchivItem_BitmapBase.cpp:
```diff
@@ -5,7 +5,12 @@
+#include "SDL/SDL.h"
+#include "SDL/SDL_opengl.h"
+#else
 #include <glad/glad.h>
```
The Handshake is quite simple:
- Server sends player id to client on connect: GameMessage_Player_Id https://github.com/Return-To-The-Roots/s25client/blob/master/libs/s25main/network/GameServer.cpp#L848
- Client answers with own version: https://github.com/Return-To-The-Roots/s25client/blob/master/libs/s25main/network/GameClient.cpp#L417
- Server confirms the version: https://github.com/Return-To-The-Roots/s25client/blob/master/libs/s25main/network/GameServer.cpp#L929

ololoken commented:
> It seems like you have the same problem with the graphics/shadows as my Android port of rttr. Do you know why this happens? Would help me a lot :D

There must something related to textures format/internal format. I'm looking at TerrainRenderer and something tells me that 99% of its code can be done by shaders.

ololoken commented:
Well, if I had planning to renderer for terrain there must be one big texture atlas with terrain sprites, generated texture with tiles indexes from atlas + gl program which accepts both 2d samples and by pixel copies atlas sprites to it's onscreen positions using mapping from indexes texture. Then in same way apply gl program to render dither between terrain connections water/land greenland/winter etc. Then gl program to render roads, then objects. 
Not sure when, but I will try to implement it.

Farmer-Markus commented:
> > It seems like you have the same problem with the graphics/shadows as my Android port of rttr. Do you know why this happens? Would help me a lot :D
> 
> There must something related to textures format/internal format. I'm looking at TerrainRenderer and something tells me that 99% of its code can be done by shaders.

I've fixed the shading issues. I don't know if this works for your web port but here are the changes(just the few in the terrain renderer)

https://github.com/Farmer-Markus/s25rttr-android/blob/main/patch%2Fs25client.patch#L2207

ololoken commented:
What I'm working on: 
```glsl
precision highp float;
attribute vec2 aVertexPosition;

uniform mat3 projectionMatrix;

varying vec2 vTextureCoord;

uniform vec4 inputSize;
uniform vec4 outputFrame;

vec4 filterVertexPosition (void) {
    vec2 position = aVertexPosition * max(outputFrame.zw, vec2(0.)) + outputFrame.xy;

    return vec4((projectionMatrix * vec3(position, 1.)).xy, 0., 1.);
}

vec2 filterTextureCoord (void) {
    return aVertexPosition * (outputFrame.zw * inputSize.zw);
}

void main (void) {
    gl_Position = filterVertexPosition();
    vTextureCoord = filterTextureCoord(); // get texture coord
}
```
and main magic happens in fragment shader

```glsl
precision highp float;
varying vec2 vTextureCoord;
uniform vec4 outputFrame;
uniform vec4 inputSize;

uniform sampler2D uAtlas; // really big texture with all loaded sprites
uniform vec2 uAtlasSize; // it's size
uniform vec2 uTileSize; // for now supports only fixed tile size
uniform vec2 uOutSize;// "real" output size

uniform sampler2D uMapping; // texture where we store in "zw" vec2 coords of sprite in atlas
uniform vec2 uMappingSize; // mapping texture size
uniform vec2 uBackgroundSize; // background size in map points
uniform vec2 uBackgroundOffset; // background offset in pixels

void main () {
    // copy current pixel color from atlas using mapping texture

    vec2 uv = vTextureCoord/outputFrame.zw*inputSize.xy;//remap to [0..1] as output depends on "inner" tileset
    vec2 tileAtlas = uTileSize/uAtlasSize;
    vec2 screenTile = (uv*uOutSize+uBackgroundOffset)/uTileSize;

    vec2 tileCoord = floor(screenTile); //  index in mapping sample 
    vec2 tileInnerCoord = fract(screenTile)*tileAtlas; // relative coords

    float mappingIndex = tileCoord.y*uBackgroundSize.x+tileCoord.x;
    float mappingY = floor(mappingIndex/uMappingSize.x);
    float mappingX = mappingIndex-mappingY*uMappingSize.x;
    vec2 shapeCoord = vec2(mappingX, mappingY)/uMappingSize;
    // here we finally found x,y of sprite in atlas texture 
    vec2 tileAtlasCoord = 255.0*texture2D(uMapping, shapeCoord).xy*tileAtlas;
    // finally output pixel color using 
    gl_FragColor = texture2D(uAtlas, tileAtlasCoord+tileInnerCoord);
}
```

nothing really to show yet, but I fill that it might and will work.

ololoken commented:
<img width="1201" alt="image" src="https://github.com/user-attachments/assets/d1a17b93-8eba-446b-852d-1b03b6f2a2d8" />
Looks better, runs faster. Will perform some cleanups, add save/load functionality and update PR

Spikeone commented:
> > Looks better, runs faster. Will perform some cleanups, add save/load functionality and update PR I think my changes will affect native app performance as well.
> 
> Great work! I also assume it will improve the native app.
> 
> Did you get the animations for water and lava working with the shader? If so can you factor out the changes to the rendering to a PR separate from the emscripten changes? Besides having more manageable PRs/changesets I'm not fully convinced about emscripten. So for me that would depend on how intrusive the changes required are and of course what @Flow86 thinks about it.
> 
> But as mentioned: I'd really love to have a shader based renderer in any case

I'd love shader based renderer, maybe that'd help improving the lightning and would be closer to the original :)

ololoken commented:
ok, maybe my update will be a bit disappointing, but I overestimated my dev capabilities to impose such significant refactoring like rendering via gl program. 
The good stuff is that I noticed bottleneck in rendering loop while I was trying. And replacement of glDrawArrays with glBegin/glEnd did the magic.

Flamefire commented:
> The good stuff is that I noticed bottleneck in rendering loop while I was trying. And replacement of glDrawArrays with glBegin/glEnd did the magic. Minimum of changes with great result, I like it.

Sounds great! What exactly was that bottleneck?

As for that replacement: I'm actually surprised by that. The immediate mode (`glBegin/glEnd`) puts more stress on the CPU with the additional driver calls and is deprecated in OpenGL 3 and removed in OpenGL ES and partially(?) in OGL 3.1   
I have found reports that `glDrawArrays` was not faster than immediate mode with explanations that this likely happens in environments doing software rendering, i.e. processing of the OpenGL pipeline is done partially/mostly on CPU so you don't gain much by batching the calls.

Especially as we have users using OpenGL ES we can't switch back to immediate mode.

However having this identified as a bottleneck is still very useful as it means we can improve our usage in this area, e.g. by not transmitting the data on every call but storing it server-side/on GPU.   
We should benchmark the difference and see where / which objects benefit most of the change and see how to optimize those using the new API.

Flamefire commented:
I guess we need to check if this is the same for the native builds or if the difference isn't there or opposite of the emscripten build.   

Maybe there is some emulation of the OpenGL 3 APIs in the library onto the immediate mode pipeline?

ololoken commented:
Definitely there is emulation https://github.com/ptitSeb/gl4es/blob/a744af14d4afbda77bf472bc53f43b9ceba39cc0/src/gl/gl4es.c#L246

ololoken commented:
@midzer did you try https://turch.in/rttr/index.html ?
If something goes wrong try to clear indexeddb cache: dev console -> application tab -> storage -> indexeddb -> /settlers2 db.
Yep I forgot to add patch for s25client.js.
```patch
Index: src/assets/module/s25client.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/assets/module/s25client.js b/src/assets/module/s25client.js
--- a/src/assets/module/s25client.js
+++ b/src/assets/module/s25client.js	(date 1763978741425)
@@ -2947,6 +2947,14 @@
                                     var parts = addr.split("/");
                                     url = url + parts[0] + ":" + port + "/" + parts.slice(1).join("/")
                                 }
+                                else {
+                                    url = (u => {
+                                        u.searchParams.append('type', sock.type);
+                                        u.searchParams.append('addr', addr);
+                                        u.searchParams.append('port', port);
+                                        return u;
+                                    })(new URL(url))
+                                }
                                 if (subProtocols !== "null") {
                                     subProtocols = subProtocols.replace(/^ +| +$/g, "").split(/ *, */);
                                     opts = subProtocols
@@ -2955,7 +2963,18 @@
                                 {
                                     WebSocketConstructor = WebSocket
                                 }
-                                ws = new WebSocketConstructor(url, opts);
+                                if (addr === '127.0.0.1' || addr === 'localhost') {
+                                    ws = {
+                                        readyState: WebSocket.OPEN,
+                                        OPEN: WebSocket.OPEN,
+                                        CLOSED: WebSocket.CLOSED,
+                                        CLOSING: WebSocket.CLOSING,
+                                        CONNECTING: WebSocket.CONNECTING,
+                                    }
+                                }
+                                else {
+                                    ws = new WebSocketConstructor(url);
+                                }
                                 ws.binaryType = "arraybuffer"
                             } catch (e) {
                                 throw new FS.ErrnoError(23)
@@ -3125,8 +3144,34 @@
                         var peer = SOCKFS.websocket_sock_ops.createPeer(sock, addr, port);
                         sock.daddr = peer.addr;
                         sock.dport = peer.port;
-                        sock.connecting = true
+                        sock.connecting = true;
+
+                        if (peer.addr === '127.0.0.1' || peer.addr === 'localhost') {
+                            var newsock = SOCKFS.createSocket(window.serverSock.family, window.serverSock.type, window.serverSock.protocol);
+
+                            var newpeer = SOCKFS.websocket_sock_ops.createPeer(newsock, {
+                                readyState: WebSocket.OPEN,
+                                OPEN: WebSocket.OPEN,
+                                CLOSED: WebSocket.CLOSED,
+                                CLOSING: WebSocket.CLOSING,
+                                CONNECTING: WebSocket.CONNECTING,
+                                _socket: {remoteAddress: 'localhost', remotePort: 111},
+                                send: (data) => peer.socket.onmessage({data})
+                            });
+
+                            newsock.daddr = newpeer.addr;
+                            newsock.dport = newpeer.port;
+
+                            peer.socket.send = (data) => newpeer.socket.onmessage({data});
+
+                            // push to queue for accept to pick up
+                            window.serverSock.pending.push(newsock);
+                            SOCKFS.emit('connection', newsock.stream.fd);
+                        }
                     }, listen(sock, backlog) {
+                        window.serverSock = sock;
+                        sock.server = true;
+                        return;
                         if (!ENVIRONMENT_IS_NODE) {
                             throw new FS.ErrnoError(138)
                         }

```

midzer commented:
@ololoken Thanks for your reply.

I tried your patch, but the message dialog still prints "Connecting - Please wait...". The previous error is gone though :)

Here's my link step (I manually patched the index.js <-> s25client.js afterwards)

```
em++ ../contrib/boost/lib/*.a ../contrib/lua-5.4.7/build/lib/liblua.a ../contrib/gl4es/lib/libGL.a libminiupnpc.a libsamplerate.a extras/*/*/*.a libs/*/*.a external/*/*/*/*.a external/*/*.a libs/s25client/CMakeFiles/s25client.dir/*.o -o index.html -sUSE_SDL=2 -sUSE_SDL_IMAGE=2 -sUSE_SDL_MIXER=2 -sSDL2_MIXER_FORMATS=ogg -sUSE_BZIP2 -fwasm-exceptions --preload-file ../data/@data/ --preload-file ../S2/@S2/ -sALLOW_MEMORY_GROWTH -sINITIAL_MEMORY=256mb -sMEMORY_GROWTH_LINEAR_STEP=128mb -sSTACK_SIZE=1024kb -sENVIRONMENT=web -sEXPORT_ALL -sEXPORTED_RUNTIME_METHODS=callMain,addRunDependency,removeRunDependency -sEXIT_RUNTIME -sFULL_ES2 -lGL -lwebsocket.js -lidbfs.js -sCASE_INSENSITIVE_FS -sWEBSOCKET_URL="ws://localhost:3665"
```

midzer commented:
@ololoken Thanks to your input, I got a version online. Don't want to share it in public though. If you want to compare our results, just message me for URL.
